scripts/opentitan/autoreg.py

```diff
@@ -0,0 +1,953 @@
+            .endianness = DEVICE_LITTLE_ENDIAN,
+            .impl = {{
+                .min_access_size = 4u,
+                .max_access_size = 4u,
+            }},
+        }};
+        '''
+        print(self.redent(code), file=tfp)
+
+    def _generate_props(self, tfp: TextIO) -> None:
+        dname = self._devname
+        hdname = dname.title().replace('_', '')
+        code = f'''
+        static Property {dname}_properties[] = {{
+            DEFINE_PROP_STRING(OT_COMMON_DEV_ID, {hdname}State, ot_id),
+            DEFINE_PROP_END_OF_LIST(),
+        }};
+        '''
+        print(self.redent(code), file=tfp)
+
+    def _generate_reset_enter(self, tfp: TextIO) -> None:
+        dname = self._devname
+        hdname = dname.title().replace('_', '')
+        uname = dname.upper()
+        code = f'''
+        static void {dname}_reset_enter(Object *obj, ResetType type)
+        {{
+            {hdname}Class *c = {uname}_GET_CLASS(obj);
+            {hdname}State *s = {uname}(obj);
+
+            if (c->parent_phases.enter) {{
+                c->parent_phases.enter(obj, type);
+            }}
+
+            memset(s->regs, 0, REG_SIZE);
+        '''
+        print(self.redent(code), file=tfp)
+        nibcount = self._regwidth // 4
+        for reg in self._registers:
+            self._generate_reset(reg, nibcount, tfp)
+        print('}\n', file=tfp)
+
+    def _generate_reset_hold(self, tfp: TextIO) -> None:
+        dname = self._devname
+        hdname = dname.title().replace('_', '')
+        uname = dname.upper()
+        code = f'''
+        static void {dname}_reset_hold(Object *obj, ResetType type)
+        {{
+            {hdname}Class *c = {uname}_GET_CLASS(obj);
+            {hdname}State *s = {uname}(obj);
+
+            if (c->parent_phases.exit) {{
+                c->parent_phases.exit(obj, type);
+            }}
+        }}
+        '''
+        print(self.redent(code), file=tfp)
+
+    def _generate_reset_exit(self, tfp: TextIO) -> None:
+        dname = self._devname
+        hdname = dname.title().replace('_', '')
+        uname = dname.upper()
+        code = f'''
+        static void {dname}_reset_exit(Object *obj, ResetType type)
+        {{
+            {hdname}Class *c = {uname}_GET_CLASS(obj);
+            {hdname}State *s = {uname}(obj);
+
+            if (c->parent_phases.exit) {{
+                c->parent_phases.exit(obj, type);
+            }}
+        }}
+        '''
+        print(self.redent(code), file=tfp)
+
+    def _generate_realize(self, tfp: TextIO) -> None:
+        dname =self._devname
+        hdname = dname.title().replace('_', '')
+        uname = dname.upper()
+        code = f'''
+        static void {dname}_realize(DeviceState *dev, Error **errp)
+        {{
+            {hdname}State *s = {uname}(dev);
+            (void)errp;
+        }}
+        '''
+        print(self.redent(code), file=tfp)
+
+    def _generate_init(self, tfp: TextIO) -> None:
+        dname =self._devname
```

Comment:
Copy-pasted in many functions:
```suggestion
        dname = self._devname
```